app/scripts/lib/EntryBridge.ts

```diff
@@ -0,0 +1,201 @@
+/**
+ * The type of change that occurred.
+ */
+export type ChangeType = 'added' | 'updated' | 'deleted';
+
+/**
+ * A callback that applies a change to an entry to a destination.
+ */
+export type ApplyChangeCallback<Entry> = (type: ChangeType, entry: Entry) => void;
+
+/**
+ * A list of changes.
+ */
+type ChangeList<Entry> = Record<ChangeType, Entry[]>;
+
+/**
+ * An object containing comparison functions for entries.
+ *
+ * @param getKey - A function that returns a string key for an entry.
+ * @param areEqual - A function that returns whether two entries are equal (have the same key and value).
+ */
+export type EntryComparators<Entry> = {
+  getKey: (entry: Entry) => string;
+  areEqual: (a: Entry, b: Entry) => boolean;
+};
+
+/**
+ * Generates a list of changes between two arrays of entries.
+ *
+ * @param prevEntries - The previous array of entries.
+ * @param newEntries - The new array of entries.
+ * @param comparators - The object containing comparison functions.
+ * @returns An object categorizing the changes into added, updated, and deleted.
+ */
+function computeChangeList<Entry>(
+  prevEntries: Entry[],
+  newEntries: Entry[],
+  comparators: EntryComparators<Entry>,
+): ChangeList<Entry> {
+  const { getKey, areEqual } = comparators;
+  const added: Entry[] = [];
+  const updated: Entry[] = [];
+  const deleted: Entry[] = [];
+
+  const prevEntriesMap = new Map(prevEntries.map((e) => [getKey(e), e]));
+  const newEntriesMap = new Map(newEntries.map((e) => [getKey(e), e]));
+
+  newEntriesMap.forEach((newEntry, newKey) => {
+    const oldEntry = prevEntriesMap.get(newKey);
+    if (oldEntry) {
+      if (!areEqual(newEntry, oldEntry)) {
+        updated.push(newEntry);
+      }
+    } else {
+      added.push(newEntry);
+    }
+  });
+
+  prevEntriesMap.forEach((oldEntry, oldKey) => {
+    if (!newEntriesMap.has(oldKey)) {
+      deleted.push(oldEntry);
+    }
+  });
+
+  return { added, updated, deleted };
+}
+
+/**
+ * Applies the changes from a ChangeList to a given function.
+ *
+ * @param changeList - The list of changes.
+ * @param applyChange - The function to apply changes.
+ */
+function applyChangeList<Entry>(
+  changeList: ChangeList<Entry>,
+  applyChange: ApplyChangeCallback<Entry>,
+): void {
+  const { added, updated, deleted } = changeList;
+  added.forEach((entry) => applyChange('added', entry));
+  updated.forEach((entry) => applyChange('updated', entry));
+  deleted.forEach((entry) => applyChange('deleted', entry));
+}
+
+/**
+ * Configuration for a one-way bridge between entries in a source and entries
+ * in a destination.
+ *
+ * @param comparators - The object containing comparison functions.
+ * @param getSourceEntries - A function that returns the entries from the
+ * source.
+ * @param applyChangeToDestination - A function that applies a change to the
+ * destination.
+ */
+export type OneWayBridgeConfig<Entry> = {
```

Comment:
I'm really impressed with what you've done here by abstracting the concept of a bridge itself, but I fear that at this level of abstraction that there is a trade off in additional complexity and code plus I worry it's unlikely other use cases would re-use this given it's still specific to our bridging use cases.

My intent was just for the concept of a "Petnames Bridge" to be abstracted meaning we avoid duplication between the two sources we now have (address book and account identities) but still encapsulate all our logic in petnames specific files.